### ISSUE TITLE
[BUG] Set-PnPUserProfileProperty : Access denied. You do not have permission to perform

### ISSUE THREAD

**Issue by thuld** [pnp/powershell#277]
Please see also related discussion [Can Get nut can't Set with pnp.powershell](https://github.com/pnp/powershell/discussions/248)

### Expected behavior 
Cmdlet `Set-PnPUserProfileProperty` allows to update of user-profile properties

### Actual behavior 
Error is raised:
````
Set-PnPUserProfileProperty : Access denied. You do not have permission to perform this action or access this resource.
At line:1 char:1
+ Set-PnPUserProfileProperty -Account 'test@foobar.com ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : WriteError: (:) [Set-PnPUserProfileProperty], ServerUnauthorizedAccessException
    + FullyQualifiedErrorId : EXCEPTION,PnP.PowerShell.Commands.UserProfiles.SetUserProfileProperty
````

### Steps to reproduce behavior
1. Connect SharePoint Online
2. Run following code: `Set-PnPUserProfileProperty -Account 'test@foobar.com' -PropertyName 'WorkPhone' -Value '0123456789'`

### What is the version of the Cmdlet module you are running?
`Manifest   1.3.0      PnP.PowerShell                      {Add-PnPAlert, Add-PnPApp, Add-PnPApplicationCustomizer, Add-PnPContentType...}`

### Which operating system/environment are you running PnP PowerShell on?
- [x] Windows
- [ ] Linux
- [ ] MacOS
- [ ] Azure Cloud Shell
- [ ] Azure Functions
- [ ] Other : please specify

**Comment by BaronSparky** [comment]
How did you connect to SPO? i.e. which switch on the "_Connect-PnPOnline_"?

I have noticed a similar issue this morning, i.e. "**Access denied**." when using the "_-UseWebLogin_" or "_-Interactive_" arguments. However, when I connected via "**-Credentials**" on the "_Connect-PnPOnline_", it worked.

Perhaps the issue is with the Connect-PnPOnline cmdlet.

**Comment by erwinvanhunen** [comment]
The issue is most likely connected to the registration of the PnP Management Shell application that is in place in your Azure AD. Run `Register-PnPManagementShellApplication` again, it will change the granted permissions: we added the userprofile readwrite right there like 2 weeks ago (it used to be only read access).

**Comment by thuld** [comment]
I am using the following approach to connect to SharePoint Online:
````
Connect-PnPOnline -Url 'https://alfapeople4com-admin.sharepoint.com' -Interactive
````

@erwinvanhunen We registered the application yesterday and the permissions of this application seems ok:

![image](https://user-images.githubusercontent.com/1762830/108037119-6e3efa80-7039-11eb-8c99-f9c2c2d70830.png)

**Comment by BaronSparky** [comment]
The same for me also...

**Comment by erwinvanhunen** [comment]
I just tested and indeed it doesn't work with a bearer token (which is what we use by default). It seems that that is (not more?) supported. We'll investigate that. I noticed that if you use -UseWebLogin (which is cookie based auth) it does work.

I'll leave this issue open while we investigate.

**Comment by Geo-Ron** [comment]
@erwinvanhunen Is there any progress on this?

**Comment by SebasT87** [comment]
Same problem here, hoping for a quick fix

**Comment by barkerboy8** [comment]
I'm having the same issue - when will this be fixed?

**Comment by Markarend** [comment]
My customer needs a fix for this!  Here are a couple of possible clues, though I still get the same error behavior discussed in this issue.

[Accessing SharePoint using an application context, also known as app-only](https://docs.microsoft.com/en-us/sharepoint/dev/solution-guidance/security-apponly) says "User Profile CSOM write operations do not work with Azure AD application - read operations work. Both read and write operations work through SharePoint App-Only principal". Then [Granting access using SharePoint App-Only](https://docs.microsoft.com/en-us/sharepoint/dev/solution-guidance/security-apponly-azureacs) gives instructions for setting that up. I was feeling pretty good when I found this, but then when I tested, still "Access denied. You do not have permission to perform this action or access this resource. Microsoft.SharePoint.Client.ServerUnauthorizedAccessException, ServerErrorCode : -2147024891, ServerErrorTypeName : System.UnauthorizedAccessException".

**Another possible clue** is that [Granting access using SharePoint App-Only](https://docs.microsoft.com/en-us/sharepoint/dev/solution-guidance/security-apponly-azureacs) shows **-AppId** and **-AppSecret** parameters for Connect-PnPOnline, but the current version (1.3.0?) doesn't provide those parameters, only -ClientId and -ClientSecret.  

_Is there a PnP core version I can test with that provides -AppId and -AppSecret?_

**Comment by Markarend** [comment]
@SebasT87 what version of SharePointPnP are you using as a workaround?

**Comment by Markarend** [comment]
@veronicageek, my apologies for seeming demanding, the exclamation point is a bit overused these days!  I absolutely appreciate everyone's supporting each other, and I'll post something if I find an answer. I'm trying some things in this article now: https://dev.to/svarukala/introducing-the-new-pnp-powershell-based-on-net-core-3-1-and-learn-how-it-s-authentication-works-pn7.

**Comment by Markarend** [comment]
Can anyone share what **Connect-PnPOnline** parameter set and values to use to authenticate from an Azure Function App so the script can _write_ User Profile properties without requiring an interactive login?

I used **Register-PnPManagementShellAccess** successfully and it appears to be configured correctly. But I'm unsure how to tell Connect-PnPOnline to use the **PnP Management Shell** app that it installs, except with -Interactive which only works locally, not from an Azure Function App.

Also tried these approaches to no avail. All can read SP profile properties but none can write:
[Connect-PnPOnline docs Example 6](https://docs.microsoft.com/en-us/powershell/module/sharepoint-pnp/connect-pnponline?view=sharepoint-ps#example-6) (why doesn't this work if the app has SharePoint | User.ReadWrite.All | Application | Read and write user profiles?)

[Granting access using SharePoint App-Only](https://docs.microsoft.com/en-us/sharepoint/dev/solution-guidance/security-apponly-azureacs) (may be deprecated, but moot because it doesn't work though it says it should for this specific scenario)

Many thanks

**Comment by Markarend** [comment]
Finally found a way to authenticate to SharePoint online to _write_ profile properties without an interactive/user logon, that works from Azure function app.  Unfortunately it uses ACS which is retired now but [_explicitly still supported for SharePoint use_](https://docs.microsoft.com/en-us/sharepoint/dev/solution-guidance/security-apponly-azureacs). So it's not my preferred approach, but as it actually works, it's better than all other approaches so far.

The key is to follow this article [Granting access using SharePoint App-Only](https://docs.microsoft.com/en-us/sharepoint/dev/solution-guidance/security-apponly-azureacs), but to add the scope /sharepoint/**social**/tenant.  Update the Permission Request XML as follows, and it will permit writing user profile properties.
```
<AppPermissionRequests AllowAppOnlyPolicy="true">
  <AppPermissionRequest Scope="http://sharepoint/content/tenant" Right="FullControl" />
  <AppPermissionRequest Scope="http://sharepoint/social/tenant" Right="FullControl" />
</AppPermissionRequests>
```
Still looking for a more "modern" approach that's not just supported as a waiver from a retired approach.  Many thanks to all!

**Comment by patrickTimmerman** [comment]
Hello, Is there an ETA on this Bug when it will be resolved ?

**Comment by ToddKlindt** [comment]
> Hello, Is there an ETA on this Bug when it will be resolved ?

Like @veronicageek said upthread, all of this work is done by volunteers in their spare time. There usually isn't an ETA and some bugs don't get fixed.

**Comment by Geo-Ron** [comment]
> @Markarend
> 
> > My customer needs a fix for this!
> 
> I'd like to remind you that this repo is community driven, from people contributing on their own personal time. So please be patient as we may also have unforeseen priorities.

This is something I was unaware of.

**Comment by erwinvanhunen** [comment]
To give you an update: this is not an issue with PnP PowerShell but has to do with how SharePoint Online handles authorization. We provided Microsoft with this feedback and are waiting for them to reply on it. We have no ETA when that will happen.

**Comment by mirontoli** [comment]
> I just tested and indeed it doesn't work with a bearer token (which is what we use by default). It seems that that is not (more?) supported. We'll investigate that. I noticed that if you use -UseWebLogin (which is cookie based auth) it does work.
> 
> I'll leave this issue open while we investigate.

I tested it with `-WebLogin` and it indeed worked, thank you for that workaround. My environment is Windows, PnP.PowerShell 1.5.0.

**Comment by wimp777** [comment]
This thread was closed? The bug is still active. With -interactive or using an Azure AD registered application with full control of sites and read/write of user profiles the error still exists.

**Comment by gautamdsheth** [comment]
Bot closed this issue due to inactivity, have opened it. The best option now would be to contact MS support and explain the case to them.

**Comment by wimp777** [comment]
Right. Which they will point back to this thread as its a community driven module and not maintained by Microsoft. I'll probably fall back to the old Sharepoint module and hope PnP gets updated as a work around.

**Comment by gautamdsheth** [comment]
You can do a minimal CSOM code and use the same access token for the support case. The underlying code is all CSOM C#. Unfortunately we have no clue if and when it will be fixed at all. Other cmdlets work perfectly fine with this auth flow but somehow this particular one doesn’t.

**Comment by Markarend** [comment]
@wimp777 and as mentioned above the workaround for Azure AD registered application is to use [Granting access using SharePoint App-Only](https://docs.microsoft.com/en-us/sharepoint/dev/solution-guidance/security-apponly-azureacs), which works if you also add the scope **http://sharepoint/social/tenant** to those instructions.  I know it's not the same as Azure AD registered but it's close and it works.

**Comment by veronicageek** [comment]
As @erwinvanhunen mentioned earlier in this thread, this isn't an issue with PowerShell PnP but dependant on MSFT's actions.

>To give you an update: this is not an issue with PnP PowerShell but has to do with how SharePoint Online handles authorization. We provided Microsoft with this feedback and are waiting for them to reply on it. We have no ETA when that will happen.

**Comment by EA12** [comment]
For me it is also working with -UseWebLogin

**Comment by erwinvanhunen** [comment]
Unfortunately the SharePoint Online support engineers are not always 100% up to date about the limitations of their own products... :-( In this case the issue lies with the fact that a requested permission through an Azure AD application grant is not being recognized by SharePoint online specifically when it comes to the UPS (User Profile Service Application). With -UseWebLogin we use a different type of authentication (cookiebased/legacy) where as with all the other authentication methods we use the currently by Microsoft promoted way of authentication : Azure AD OAuth. We switched over to Azure AD OAuth as the default as more and more API requests we need to make are located in the Microsoft Graph. And the MS Graph only supports Azure AD OAuth. 

In order for the OAuth authentication flow to work one needs to grant permissions to an Azure AD App registration, either a multi-tenant one alike the PnP Management Shell, or your own registered one. Our tests show that while the correct permissions are in place, SPO is not allowing the user to perform the actions. We brought this up with the product manager in charge of those APIs back then but are still awaiting feedback.

**Comment by KoenZomers** [comment]
> @veronicageek I opened a case with Microsoft and escalated this with their SharePoint Online engineers for this issue. I was told that this is not an issue on their end and that the issue is with the module itself.

Please share your case number so I can look into this. Either here or send it to me privately.

**Comment by apinales** [comment]
25392520

Avel Pinales
Senior Consultant
P 216.359.7866 | E ***@***.******@***.***>

***@***.***D761D1.DF053AC0]<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fcorebts.com%2F&data=04%7C01%7CAvel.Pinales%40COREBTS.com%7Cb503bfa7e81546028c4208d8cd0bebf6%7Cb389dae6fcaf4bd79d457e96e87d725e%7C0%7C0%7C637484798385484649%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=1bQ7cC3%2B5MhYO0ROlPEYyMPXKd8x8iDZnQxg%2BIR4nfg%3D&reserved=0>

**Comment by KoenZomers** [comment]
Thanks for sharing @apinales. Looking into it with the support team.

**Comment by KoenZomers** [comment]
@apinales Discussed this with the support team. We would very much like to dive into this issue but since your case has been closed two months ago already we would like to ask you to open a new case and have it reference your old case. If you can share the new case number with me, we'll ensure the correct details will be transferred to the engineer who will handle the case.

**Comment by KoenZomers** [comment]
Thanks for creating the ticket @apinales. I'm already aware of the ticket number, no need to share.

**Comment by mohandivraniya** [comment]
I am running into this issue right now. Some updates:

- It was working fine with SharePointPnPPowerShellOnline fine, issue started appearing when we migrated to PnP.PowerShell 1.5.0.
- Raised a case with Microsoft support and they are looking into it.
- Got some updates from them to upgrade to 1.6.0 to see if that resolves the issue. However, it does not. Facing same issue with PnP.PowerShell version 1.6.0
- Tried -UseWebLogin and it does work. -Interactive and -Credentials do not work. However, I can't really use -UseWebLogin since we are running this script from an Azure automation account runbook.
- Re-ran Register-PnPManagementShellAccess and provided global admin consent again, that didn't change outcome. 
- I don't get "Access Denied" error text. For me, the error message is : "This operation requires you to be managing your own data or have administrator privileges." I tried this with SharePoint administrator and even a global administrator role role, however outcome is same.

**Comment by KoenZomers** [comment]
@mohandivraniya can you share your case number so we can bundle it with the one we already have ongoing at the moment for this?

In my repro I'm getting the exact same error message as you. What stumbles me here is that you say it works with SharePointPnPPowerShellOnline. What version of it were you using? How did you connect using that?

**Comment by mohandivraniya** [comment]
My case number - 26234370
I am using SharePointPnPPowerShellOnline version 3.23.2007.0.
I am connecting using Connect-PnPOnline -Url <<TenantAdminUrl>> -UseWebLogin

**Comment by KoenZomers** [comment]
@mohandivraniya that explains it then. WebLogin will work as it works through a cookie to authenticate. The issue I believe we are all facing here is that we cannot register an application in AAD, apply the SharePoint Application principal Users.ReadWrite.All and use that to update a user profile. You would either need to assign the same client Id social permissions through AppInv.aspx as well to make it work or alternatively use -UseWebLogin.

**Comment by Markarend** [comment]
Here's a non-interactive method for connecting that does work to change user profile properties from a script running in an Azure runbook or function: [Example 3 on this page](https://docs.microsoft.com/en-us/powershell/module/sharepoint-pnp/connect-pnponline). You must make one change. Follow this article [Granting access using SharePoint App-Only](https://docs.microsoft.com/en-us/sharepoint/dev/solution-guidance/security-apponly-azureacs), but add the scope /sharepoint/**social**/tenant.  Update the Permission Request XML as follows, and it will permit writing user profile properties.
```
<AppPermissionRequests AllowAppOnlyPolicy="true">
  <AppPermissionRequest Scope="http://sharepoint/content/tenant" Right="FullControl" />
  <AppPermissionRequest Scope="http://sharepoint/social/tenant" Right="FullControl" />
</AppPermissionRequests>
```
This ACS method is retired now but [_explicitly still supported for SharePoint use_](https://docs.microsoft.com/en-us/sharepoint/dev/solution-guidance/security-apponly-azureacs)

**Comment by KoenZomers** [comment]
Thanks @Markarend for taking the time to detail out what I mentioned above to help others better understand the workaround. Still I believe that should be seen as a workaround. We're currently trying to see if filing a hotfix request for this has any chance of succeeding.

**Comment by bcameron1231** [comment]
It would be awesome if Microsoft could fix this for Azure AD app-only contexts.  This has been an issue for a while, and you lose a bit of transparency for your app registrations when you are building AD Apps, and subsequently have to register some permissions though appinv.aspx

**Comment by heinrich-ulbricht** [comment]
We are tripping over this as well. Found this thread and it's awesome that there are workarounds and - more important - this is being worked on. 👍

Standing in line and waiting for the engineers to click "ship it" ;)

**Comment by KoenZomers** [comment]
Turns out engineering already had a backlog item open to implement those permissions. The need to have this implemented is raised to engineering through a Critical Design Change Request (CDCR) now based on the discussion here with the aim of prioritizing the backlog item on upcoming sprints. They're currently looking into the efforts it would take to implement this.

**Comment by KoenZomers** [comment]
Just as an update, the topic is still on the table and has not been forgotten. No decision has been made yet if this functionality will be adjusted or not unfortunately.

**Comment by asfd79** [comment]
Any update from Microsoft SharePoint Online on this issue?????

**Comment by KoenZomers** [comment]
> Any update from Microsoft SharePoint Online on this issue?????

Its still under discussion. Vacation period makes things progress slow unfortunately. Also because there is an easy workaround available, it takes more effort to raise the priority on this.

**Comment by asfd79** [comment]
@KoenZomers Thanks for updating me. The easy workaround you are talking about is app-only credential?

**Comment by KoenZomers** [comment]
> @KoenZomers Thanks for updating me. The easy workaround you are talking about is app-only credential?

The easy workaround is to apply what is written here:
https://github.com/pnp/powershell/issues/277#issuecomment-866839692

So you can combine your AAD App Registration with ACS permissions to make this work. You would just need to get another token from ACS to handle the user profile stuff, if you directly communicate with the SharePoint/Graph APIs. If you use PnP PowerShell, this will be done automatically for you and you just need to set the permissions as mentioned in the referenced post.

**Comment by asfd79** [comment]
@KoenZomers Thanks, mate. I give it a go. Really really hope Microsoft fix it properly so.

**Comment by brianpmccullough** [comment]
I'd like to add that this is something I'd like to see resolved for a current client.  While I can (barely) keep up with the subtle nuance and intricacies of things like this, clients/customers often times aren't paying as close attention.  As a result those that have to build/deliver/transition solutions that require both Azure App Registration and the legacy ACS permissions have a hard time explaining this to customer teams.  It also requires delivery of code that will be known to have a limited life span in that ACS will only likely be supported for some (unknown) period of time.  I'd like to vote this up in priority / escalation such that administrators can have custom SPO user profile sync jobs that don't require ACS any longer.

**Comment by jeffangama** [comment]
Hi,
I tried the workaround method suggested using SharePoint-App Online registration, and using this function of pnp powershell code : Set-PnPUserProfileProperty
I am now getting error 401, did i miss something ?

**Comment by Markarend** [comment]
@jeffangama did you add an AppPermissionsRequest as described here

When using [SharePoint App-Only](https://docs.microsoft.com/en-us/sharepoint/dev/solution-guidance/security-apponly-azureacs), must add the scope /sharepoint/social/tenant. Update the Permission Request XML as follows, and it will permit writing user profile properties.

<AppPermissionRequests AllowAppOnlyPolicy="true">
  <AppPermissionRequest Scope="http://sharepoint/content/tenant" Right="FullControl" />
  <AppPermissionRequest Scope="http://sharepoint/social/tenant" Right="FullControl" />
</AppPermissionRequests>

**Comment by MartijnMolegraaf** [comment]
@Markarend thx for you contribution to this topic. Very helpfull ! I was using the [SharePoint App-Only](https://docs.microsoft.com/en-us/sharepoint/dev/solution-guidance/security-apponly-azureacs) method already. It works at our test tenant, but at the customer it gives an error.

I used exactly the same XML as you do at /_layouts/15/appinv.aspx
```
<AppPermissionRequests AllowAppOnlyPolicy="true">
<AppPermissionRequest Scope="http://sharepoint/social/tenant" Right="FullControl" />
<AppPermissionRequest Scope="http://sharepoint/content/tenant" Right="FullControl" />
</AppPermissionRequests>
```

I used this code:

```
$GLOBAL:spAppOnlyClientId = "XXXX"
$GLOBAL:spAppOnlyClientSecret = "XXXX"
$GLOBAL:connection = Connect-PnPOnline -url "https://<tenant>-admin.sharepoint.com" -clientId $GLOBAL:spAppOnlyClientId -clientSecret $GLOBAL:spAppOnlyClientSecret
(Get-PnPUserProfileProperty -Account "<upn>" -Connection $GLOBAL:connection).UserProfileProperties
```

The error I receive is:
`The remote server returned an error: (401) Unauthorized.`

Do you have any clue why this isn't working?

As an alternative I tried the -UseWebLogin method, which works. But it should run in a Azure runbook, so this method is of no use to us.

**Comment by KoenZomers** [comment]
@MartijnMolegraaf Check if ACS auth is disabled on the tenant where it gives you the unauthorized.

```powershell
Connect-SPOService -Url https://tenant-admin.sharepoint.com
Get-SPOTenant | select DisableCustomAppAuthentication
```

Ensure it reports back False. If not, you need to set it to false using:

```powershell
Set-SPOTenant -DisableCustomAppAuthentication $false
```

**Comment by KoenZomers** [comment]
As an update to this item. We have discussed the situation with engineering. They are aware of this situation and this GitHub thread and are evaluating their options.

**Comment by MartijnMolegraaf** [comment]
Morning @KoenZomers . You made my day, that was indeed the cause and solution. Thank you !

In addition, it is appreciated that you have discussed this case and looked at possible options.

**Comment by maartendamen** [comment]
Any news on this issue?

**Comment by KoenZomers** [comment]
Not yet unfortunately. Its not getting priority yet to be looked at. We'll keep pushing.

**Comment by veronicageek** [comment]
@KoenZomers - When the problem is fixed from MSFT side, will it be automatically updated in the PnP module or will we have to also fix our code?

**Comment by KoenZomers** [comment]
@veronicageek depends a bit on how they decide to fix it. Most probably it will not need code changes from our end.

**Comment by maartendamen** [comment]
What would currently be the best way to update SharePoint profile photos despite of this bug?

**Comment by warwickofthegh** [comment]
> What would currently be the best way to update SharePoint profile photos despite of this bug?

Utilising New-PnPUPABulkImportJob achieves the same result in a roundabout way

**Comment by maartendamen** [comment]
> > What would currently be the best way to update SharePoint profile photos despite of this bug?
> 
> Utilising New-PnPUPABulkImportJob achieves the same result in a roundabout way

Would you mind sharing a basic example?

**Comment by KoenZomers** [comment]
@maartendamen have a look at the documentation: https://pnp.github.io/powershell/cmdlets/New-PnPUPABulkImportJob.html . The idea is that you write the updates you want to do in a JSON file, upload it somewhere to SharePoint Online and then call this cmdlet to ask to process it for you. Alternatively you can also directly connect Azure Active Directory with the SharePoint Online User Profile store using https://pnp.github.io/powershell/cmdlets/Sync-PnPSharePointUserProfilesFromAzureActiveDirectory.html.

Permissions wise you will not solve anything with this by the way.

**Comment by JohnKoornneef** [comment]
Is there any way to upvote this issue at Microsoft? Because the old module SharePointPnPPowerShellOnline still works for me. But it seems that it will be retired. And the new method doesn't work. That doesn't feel right for a production environment.

**Comment by KoenZomers** [comment]
I know this one has been open for quite some time already and that there are a lot of organizations that are hitting this issue. I can say that we're closer than ever on getting this addressed. I'm not at liberty to give any timelines on it, but know that we are still actively working on it internally to get it resolved. Thank you for your patience everyone.

**Comment by KoenZomers** [comment]
> Is there any way to upvote this issue at Microsoft? Because the old module SharePointPnPPowerShellOnline still works for me. But it seems that it will be retired. And the new method doesn't work. That doesn't feel right for a production environment.

No need to. It has priority already.

You're okay with using the SharePointPnPPowerShellOnline library for now. If it works, it works. It will likely also still work for you with PnP.PowerShell as long as you use the same login method, i.e. -UseWeblogin. Nothing has changed with regards to that.

**Comment by Yvand** [comment]
Everyone - I'm working at Microsoft Support, and with @KoenZomers and Engineering on this issue.
Microsoft Engineering has analyzed this issue and is currently working on a fix. It requires collaboration between multiple teams and there is no ETA for the moment. I will update this thread as updates occur

**Comment by vvdadhe** [comment]
I tested using  SharePointPnPPowerShellOnline  legacy module and it works at my end, although it is older module than PnP.PowerShell.

**Comment by brianpmccullough** [comment]
Still patiently waiting on Microsoft to fix this (not the community).

**Comment by silentmark** [comment]
I will give few points here as well, as i run into similar problem, but only when the script is running as Azure Runbook. When the script is running locally (in same windows powershell 5.1 version) it works well. 

So - connecting with ClientID and Secret gives warning, allows to get property and update property. 
when run from Azure Runbook, throws error: 

`at <ScriptBlock>, <No file>: line 68



PSMessageDetails      : 
Exception             : System.Reflection.TargetInvocationException: Exception has been thrown by the target of an 
                        invocation. ---> System.Exception: Token request failed. ---> System.Net.WebException: The 
                        remote server returned an error: (401) Unauthorized.
                           at System.Net.HttpWebRequest.GetResponse()
                           at SharePointPnP.IdentityModel.Extensions.S2S.Protocols.OAuth2.OAuth2WebRequest.GetResponse()
                           at SharePointPnP.IdentityModel.Extensions.S2S.Protocols.OAuth2.OAuth2S2SClient.Issue(String 
                        securityTokenServiceUrl, OAuth2AccessTokenRequest oauth2Request)
                           --- End of inner exception stack trace ---
                           at SharePointPnP.IdentityModel.Extensions.S2S.Protocols.OAuth2.OAuth2S2SClient.Issue(String 
                        securityTokenServiceUrl, OAuth2AccessTokenRequest oauth2Request)
                           at PnP.Framework.Utilities.TokenHelper.GetAppOnlyAccessToken(String targetPrincipalName, 
                        String targetHost, String targetRealm)
                           at PnP.Framework.Utilities.ACSTokenGenerator.GetToken(Uri siteUrl)
                           at 
                        PnP.Framework.AuthenticationManager.<>c__DisplayClass75_0.<GetAccessTokenContext>b__0(Object 
                        sender, WebRequestEventArgs args)
                           at 
                        Microsoft.SharePoint.Client.ClientRuntimeContext.OnExecutingWebRequest(WebRequestEventArgs args)
                           --- End of inner exception stack trace ---
                           at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, 
                        Boolean constructor)
                           at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, 
                        Object[] arguments)
                           at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder 
                        binder, Object[] parameters, CultureInfo culture)
                           at 
                        Microsoft.SharePoint.Client.ClientContextExtensions.<>c__DisplayClass10_0.<Clone>b__0(Object 
                        oSender, WebRequestEventArgs webRequestEventArgs)
                           at System.EventHandler`1.Invoke(Object sender, TEventArgs e)
                           at 
                        Microsoft.SharePoint.Client.ClientRuntimeContext.OnExecutingWebRequest(WebRequestEventArgs args)
                           at Microsoft.SharePoint.Client.ClientContext.GetWebRequestExecutor()
                           at Microsoft.SharePoint.Client.ClientContext.<GetFormDigestInfoPrivateAsync>d__37.MoveNext()
                        --- End of stack trace from previous location where exception was thrown ---
                           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
                           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task 
                        task)
                           at Microsoft.SharePoint.Client.ClientContext.<EnsureFormDigestAsync>d__36.MoveNext()
                        --- End of stack trace from previous location where exception was thrown ---
                           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
                           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task 
                        task)
                           at Microsoft.SharePoint.Client.ClientContext.<ExecuteQueryAsync>d__28.MoveNext()
                        --- End of stack trace from previous location where exception was thrown ---
                           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
                           at 
                        Microsoft.SharePoint.Client.ClientContextExtensions.<ExecuteQueryImplementation>d__6.MoveNext()
                        --- End of stack trace from previous location where exception was thrown ---
                           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
                           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task 
                        task)
                           at 
                        Microsoft.SharePoint.Client.ClientContextExtensions.ExecuteQueryRetry(ClientRuntimeContext 
                        clientContext, Int32 retryCount, String userAgent)
                           at PnP.PowerShell.Commands.UserProfiles.GetUserProfileProperty.ExecuteCmdlet()
                           at PnP.PowerShell.Commands.PnPSharePointCmdlet.ProcessRecord()
TargetObject          : 
CategoryInfo          : WriteError: (:) [Get-PnPUserProfileProperty], TargetInvocationException
FullyQualifiedErrorId : EXCEPTION,PnP.PowerShell.Commands.UserProfiles.GetUserProfileProperty
ErrorDetails          : Exception has been thrown by the target of an invocation.
InvocationInfo        : System.Management.Automation.InvocationInfo
ScriptStackTrace      : at <ScriptBlock>, <No file>: line 68
PipelineIterationInfo : {}



MyCommand             : Get-PnPUserProfileProperty
BoundParameters       : {}
UnboundArguments      : {}
ScriptLineNumber      : 68
OffsetInLine          : 23
HistoryId             : 1
ScriptName            : 
Line                  :             $props = (Get-PnPUserProfileProperty -Account 
                        $user.UserPrincipalName).UserProfileProperties
                        
PositionMessage       : At line:68 char:23
                        + ...   $props = (Get-PnPUserProfileProperty -Account $user.UserPrincipalNa ...
                        +                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
PSScriptRoot          : 
PSCommandPath         : 
InvocationName        : Get-PnPUserProfileProperty
PipelineLength        : 0
PipelinePosition      : 0
ExpectingInput        : False
CommandOrigin         : Internal
DisplayScriptPosition : 



Message        : Exception has been thrown by the target of an invocation.
Data           : {CorrelationId, TimeStampUtc}
InnerException : System.Exception: Token request failed. ---> System.Net.WebException: The remote server returned an 
                 error: (401) Unauthorized.
                    at System.Net.HttpWebRequest.GetResponse()
                    at SharePointPnP.IdentityModel.Extensions.S2S.Protocols.OAuth2.OAuth2WebRequest.GetResponse()
                    at SharePointPnP.IdentityModel.Extensions.S2S.Protocols.OAuth2.OAuth2S2SClient.Issue(String 
                 securityTokenServiceUrl, OAuth2AccessTokenRequest oauth2Request)
                    --- End of inner exception stack trace ---
                    at SharePointPnP.IdentityModel.Extensions.S2S.Protocols.OAuth2.OAuth2S2SClient.Issue(String 
                 securityTokenServiceUrl, OAuth2AccessTokenRequest oauth2Request)
                    at PnP.Framework.Utilities.TokenHelper.GetAppOnlyAccessToken(String targetPrincipalName, String 
                 targetHost, String targetRealm)
                    at PnP.Framework.Utilities.ACSTokenGenerator.GetToken(Uri siteUrl)
                    at PnP.Framework.AuthenticationManager.<>c__DisplayClass75_0.<GetAccessTokenContext>b__0(Object 
                 sender, WebRequestEventArgs args)
                    at Microsoft.SharePoint.Client.ClientRuntimeContext.OnExecutingWebRequest(WebRequestEventArgs args)
TargetSite     : System.Object InvokeMethod(System.Object, System.Object[], System.Signature, Boolean)
StackTrace     :    at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, 
                 Boolean constructor)
                    at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, 
                 Object[] arguments)
                    at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, 
                 Object[] parameters, CultureInfo culture)
                    at Microsoft.SharePoint.Client.ClientContextExtensions.<>c__DisplayClass10_0.<Clone>b__0(Object 
                 oSender, WebRequestEventArgs webRequestEventArgs)
                    at System.EventHandler`1.Invoke(Object sender, TEventArgs e)
                    at Microsoft.SharePoint.Client.ClientRuntimeContext.OnExecutingWebRequest(WebRequestEventArgs args)
                    at Microsoft.SharePoint.Client.ClientContext.GetWebRequestExecutor()
                    at Microsoft.SharePoint.Client.ClientContext.<GetFormDigestInfoPrivateAsync>d__37.MoveNext()
                 --- End of stack trace from previous location where exception was thrown ---
                    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
                    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
                    at Microsoft.SharePoint.Client.ClientContext.<EnsureFormDigestAsync>d__36.MoveNext()
                 --- End of stack trace from previous location where exception was thrown ---
                    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
                    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
                    at Microsoft.SharePoint.Client.ClientContext.<ExecuteQueryAsync>d__28.MoveNext()
                 --- End of stack trace from previous location where exception was thrown ---
                    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
                    at Microsoft.SharePoint.Client.ClientContextExtensions.<ExecuteQueryImplementation>d__6.MoveNext()
                 --- End of stack trace from previous location where exception was thrown ---
                    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
                    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
                    at Microsoft.SharePoint.Client.ClientContextExtensions.ExecuteQueryRetry(ClientRuntimeContext 
                 clientContext, Int32 retryCount, String userAgent)
                    at PnP.PowerShell.Commands.UserProfiles.GetUserProfileProperty.ExecuteCmdlet()
                    at PnP.PowerShell.Commands.PnPSharePointCmdlet.ProcessRecord()
HelpLink       : 
Source         : mscorlib
HResult        : -2146232828

   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.SharePoint.Client.ClientContextExtensions.<>c__DisplayClass10_0.<Clone>b__0(Object oSender, WebRequestEventArgs webRequestEventArgs)
   at System.EventHandler`1.Invoke(Object sender, TEventArgs e)
   at Microsoft.SharePoint.Client.ClientRuntimeContext.OnExecutingWebRequest(WebRequestEventArgs args)
   at Microsoft.SharePoint.Client.ClientContext.GetWebRequestExecutor()
   at Microsoft.SharePoint.Client.ClientContext.<GetFormDigestInfoPrivateAsync>d__37.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.SharePoint.Client.ClientContext.<EnsureFormDigestAsync>d__36.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.SharePoint.Client.ClientContext.<ExecuteQueryAsync>d__28.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.SharePoint.Client.ClientContextExtensions.<ExecuteQueryImplementation>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.SharePoint.Client.ClientContextExtensions.ExecuteQueryRetry(ClientRuntimeContext clientContext, Int32 retryCount, String userAgent)
   at PnP.PowerShell.Commands.UserProfiles.GetUserProfileProperty.ExecuteCmdlet()
   at PnP.PowerShell.Commands.PnPSharePointCmdlet.ProcessRecord()



Message        : Token request failed.
Data           : {}
InnerException : System.Net.WebException: The remote server returned an error: (401) Unauthorized.
                    at System.Net.HttpWebRequest.GetResponse()
                    at SharePointPnP.IdentityModel.Extensions.S2S.Protocols.OAuth2.OAuth2WebRequest.GetResponse()
                    at SharePointPnP.IdentityModel.Extensions.S2S.Protocols.OAuth2.OAuth2S2SClient.Issue(String 
                 securityTokenServiceUrl, OAuth2AccessTokenRequest oauth2Request)
TargetSite     : SharePointPnP.IdentityModel.Extensions.S2S.Protocols.OAuth2.OAuth2Message Issue(System.String, 
                 SharePointPnP.IdentityModel.Extensions.S2S.Protocols.OAuth2.OAuth2AccessTokenRequest)
StackTrace     :    at SharePointPnP.IdentityModel.Extensions.S2S.Protocols.OAuth2.OAuth2S2SClient.Issue(String 
                 securityTokenServiceUrl, OAuth2AccessTokenRequest oauth2Request)
                    at PnP.Framework.Utilities.TokenHelper.GetAppOnlyAccessToken(String targetPrincipalName, String 
                 targetHost, String targetRealm)
                    at PnP.Framework.Utilities.ACSTokenGenerator.GetToken(Uri siteUrl)
                    at PnP.Framework.AuthenticationManager.<>c__DisplayClass75_0.<GetAccessTokenContext>b__0(Object 
                 sender, WebRequestEventArgs args)
                    at Microsoft.SharePoint.Client.ClientRuntimeContext.OnExecutingWebRequest(WebRequestEventArgs args)
HelpLink       : 
Source         : PnP.Framework
HResult        : -2146233088

   at SharePointPnP.IdentityModel.Extensions.S2S.Protocols.OAuth2.OAuth2S2SClient.Issue(String securityTokenServiceUrl, OAuth2AccessTokenRequest oauth2Request)
   at PnP.Framework.Utilities.TokenHelper.GetAppOnlyAccessToken(String targetPrincipalName, String targetHost, String targetRealm)
   at PnP.Framework.Utilities.ACSTokenGenerator.GetToken(Uri siteUrl)
   at PnP.Framework.AuthenticationManager.<>c__DisplayClass75_0.<GetAccessTokenContext>b__0(Object sender, WebRequestEventArgs args)
   at Microsoft.SharePoint.Client.ClientRuntimeContext.OnExecutingWebRequest(WebRequestEventArgs args)



Status         : ProtocolError
Response       : System.Net.HttpWebResponse
Message        : The remote server returned an error: (401) Unauthorized.
Data           : {}
InnerException : 
TargetSite     : System.Net.WebResponse GetResponse()
StackTrace     :    at System.Net.HttpWebRequest.GetResponse()
                    at SharePointPnP.IdentityModel.Extensions.S2S.Protocols.OAuth2.OAuth2WebRequest.GetResponse()
                    at SharePointPnP.IdentityModel.Extensions.S2S.Protocols.OAuth2.OAuth2S2SClient.Issue(String 
                 securityTokenServiceUrl, OAuth2AccessTokenRequest oauth2Request)
HelpLink       : 
Source         : System
HResult        : -2146233079

   at System.Net.HttpWebRequest.GetResponse()
   at SharePointPnP.IdentityModel.Extensions.S2S.Protocols.OAuth2.OAuth2WebRequest.GetResponse()
   at SharePointPnP.IdentityModel.Extensions.S2S.Protocols.OAuth2.OAuth2S2SClient.Issue(String securityTokenServiceUrl, OAuth2AccessTokenRequest oauth2Request)`

**Comment by parlevjo2** [comment]
Still not working with
ConnectPnPOnline -Url "myco-admin.sharepoint.com" -ClientId "ZZZZZZ" -Tenant "myco.onmicrosoft.com" -Thumbprint "XXXXXXX"

When will MSFT fix this ?

Set-PnPUserProfileProperty : This operation requires you to be managing your own data or have administrator privileges.
At line:12 char:5
+     Set-PnPUserProfileProperty -Account xxxxxx -Prope ...
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : WriteError: (:) [Set-PnPUserProfileProperty], ServerException
    + FullyQualifiedErrorId : EXCEPTION,PnP.PowerShell.Commands.UserProfiles.SetUserProfileProperty

**Comment by MudassarAli** [comment]
Waiting patiently for the fix from Microsoft.

**Comment by colonelclaypoo** [comment]
> Still patiently waiting on Microsoft to fix this (not the community).
> 
> Quick update on my end. I am able to get the "workaround" solution to work (creating Azure App Registration then using AppInv.aspx to use the legacy ACS authorization with permission scopes outlined above), but only if I Connect-PnPOnline using ClientId and ClientSecret parameters: `ConnectPnPOnline -Url "myco-admin.sharepoint.com" -ClientId "ZZZZZZ" -ClientSecret "abcdefghijklmnop" ` Note: Connect-PnPOnline gives me the warning about using legacy authentication when connecting this way. Yes, I am aware of -WarningAction Ignore.
> `

Thank you! That works for me.

**Comment by Yvand** [comment]
Everyone - I checked with Engineering a few day ago, which confirmed they are currently working on a hotfix for this issue.
The plan is to release it this calendar year, I don't have a more specific date.

**Comment by felipe-developer** [comment]
I had success using this command "Connect-PnPOnline -s -url $Url -Credentials $Creds"

**Comment by parlevjo2** [comment]
> I had success using this command "Connect-PnPOnline -s -url $Url -Credentials $Creds"

We had the same. We call it a work around, becasue we want to quit using a credential. Our line is.
Connect-PnPOnline -Url $AdminSiteURL -Credentials $credential -SPOManagementShell

**Comment by Yvand** [comment]
@WiljanD I got no update since my last message but I just asked for one, I'll post back here if/when I get an answer

**Comment by brianpmccullough** [comment]
> Everyone - I checked with Engineering a few day ago, which confirmed they are currently working on a hotfix for this issue. The plan is to release it this calendar year, I don't have a more specific date.

Time is just about up on this.  Can we expect this to be fixed anytime soon?  I have been patiently waiting for almost 2 years.

**Comment by blarrkan** [comment]
> Everyone - I checked with Engineering a few day ago, which confirmed they are currently working on a hotfix for this issue.
> The plan is to release it this calendar year, I don't have a more specific date.

Have you asked the engineer for any news on the hotfix?

**Comment by kodsu** [comment]
@Yvand Hello, have you got any news/ETA ?
For anyone having issue with non interactive login please refer to @Markarend comments which have a workaround while waiting for the Microsoft fix, especially this one https://github.com/pnp/powershell/issues/277#issuecomment-866839692

**Comment by Yvand** [comment]
@kodsu yes, I got confirmation from Engineering that they are actively working on this and they already completed a good part of the work, although it is not finished yet.
They did not communicate a timeline when it will ship in production, but it will not be before CY 2023.

**Comment by rimayug** [comment]
2023 is here. hope the teams working on this will get us a solution soon.

**Comment by mikelee1313** [comment]
Anyone updates for the new year?

**Comment by Yvand** [comment]
Today I got an update from Engineering: They continued the work and made progress since my last message, but they still don't have a more precise timeline to communicate, very unfortunately

**Comment by heinrich-ulbricht** [comment]
Imagine being a PM and getting this response from engineering ⛈️😉

**Comment by Yvand** [comment]
@heinrich-ulbricht I'm just the messenger who understands the frustration of people in this thread who are waiting for this fix for so long.
I take the time to chase Engineering and provide whatever update I have, even if it is a poor one.
I could have chosen to just remain silent

**Comment by Mntz** [comment]
Is the current work-around still working?

I created a new Azure AD App + secret.
Next on the tenant-admin.sharepoint.com/_layouts/15/appinv.aspx added the FullControl rights to both the content and social scope.
Using Connect-PnPOnline with ClientId and ClientSecret, the Get-PnPUserProfileProperty command is working fine.
But Set-PnPUserProfileProperty returns the following error:
```
Set-PnPUserProfileProperty : De huidige gebruiker is niet gemachtigd om deze bewerking uit te voeren.
At line:1 char:1
+ Set-PnPUserProfileProperty -Account $acc.AccountName -PropertyName 'S ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [Set-PnPUserProfileProperty], PSInvalidOperationException
    + FullyQualifiedErrorId : InvalidOperation,PnP.PowerShell.Commands.UserProfiles.SetUserProfileProperty
```
The error text in this Dutch tenant translates the same but the CategoryInfo differs from the topic start.

**Comment by brianpmccullough** [comment]
> Is the current work-around still working?
> 
> I created a new Azure AD App + secret. Next on the tenant-admin.sharepoint.com/_layouts/15/appinv.aspx added the FullControl rights to both the content and social scope. Using Connect-PnPOnline with ClientId and ClientSecret, the Get-PnPUserProfileProperty command is working fine. But Set-PnPUserProfileProperty returns the following error:
> 
> ```
> Set-PnPUserProfileProperty : De huidige gebruiker is niet gemachtigd om deze bewerking uit te voeren.
> At line:1 char:1
> + Set-PnPUserProfileProperty -Account $acc.AccountName -PropertyName 'S ...
> + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>     + CategoryInfo          : InvalidOperation: (:) [Set-PnPUserProfileProperty], PSInvalidOperationException
>     + FullyQualifiedErrorId : InvalidOperation,PnP.PowerShell.Commands.UserProfiles.SetUserProfileProperty
> ```
> 
> The error text in this Dutch tenant translates the same but the CategoryInfo differs from the topic start.

Did you run the following on each Geo where you need to perform the Get/Set?  Also, if multi-geo, you will need to so the ACS registration (AppInv.aspx) on each central admin site.

set-spotenant -DisableCustomAppAuthentication $false


https://learn.microsoft.com/en-us/sharepoint/dev/solution-guidance/security-apponly-azureacs

**Comment by Mntz** [comment]
Found the issue... the error was kind of misleading:
I tried to set SPS-Location but the location value was not yet available in the term store.
Since I first tried Set-PnPUserProfileProperty using certificate authentication instead of using secrets I figured it was still the same issue, but no.

**Comment by brianpmccullough** [comment]
Yes, for sync scenarios, I usually let those termsets remain open - following the model that OOB SharePoint uses for fields like Job Title, etc.

**Comment by sli701** [comment]
ERROR: This operation requires you to be managing your own data or have administrator privileges.

Getting the same error here when using Azure Function and App Registration.

Is there a fix?

**Comment by RichWorld-Tech** [comment]
Connect-PnPOnline -Url $URL -UseWebLogin. This is not ideal nor what we want for automation purposes, but it will work. When working on multiple users, I add it to the Begin block and do the work in the Process block  of the script in order to pipe information to the script. Hope that helps.

**Comment by Milkias1** [comment]
> ONE SOLUTION
> 
> Finally found a way to authenticate to SharePoint online to _write_ profile properties without an interactive/user logon, that works from Azure function app. Unfortunately it uses ACS which is retired now but [_explicitly still supported for SharePoint use_](https://docs.microsoft.com/en-us/sharepoint/dev/solution-guidance/security-apponly-azureacs). So it's not my preferred approach, but as it actually works, it's better than all other approaches so far.
> 
> The key is to follow this article [Granting access using SharePoint App-Only](https://docs.microsoft.com/en-us/sharepoint/dev/solution-guidance/security-apponly-azureacs), but to add the scope /sharepoint/**social**/tenant. Update the Permission Request XML as follows, and it will permit writing user profile properties.
> 
> ```
> <AppPermissionRequests AllowAppOnlyPolicy="true">
>   <AppPermissionRequest Scope="http://sharepoint/content/tenant" Right="FullControl" />
>   <AppPermissionRequest Scope="http://sharepoint/social/tenant" Right="FullControl" />
> </AppPermissionRequests>
> ```
> 
> Still looking for a more "modern" approach that's not just supported as a waiver from a retired approach. Many thanks to all!

Thank you, [Markarend](https://github.com/Markarend) this worked for me.

**Comment by jackpoz** [comment]
Just wanted to mention that New-PnPUPABulkImportJob works fine with an app registration with Certificate and permissions assigned only in AAD with SharePoint "Sites.FullControl.All" and "User.ReadWrite.All" and no permissions assigned in SPO at /_layouts/15/appinv.aspx , tested it this week in 2 different tenants.

**Comment by testad** [comment]
Thanks a lot @jackpoz, will definitely try this out. Glad to get rid of this ACS dependance.

**Comment by rayghost503** [comment]
> SPOM

This seems to be working for me. I just fixed an automation I had working.

**Comment by jackpoz** [comment]
I was able to test New-PnPUPABulkImportJob with Managed Identity in Automation Account in Azure and it works fine that way too

**Comment by KoenZomers** [comment]
Good to hear the feedback here that it has started to work for some of you. Just for awareness, the PG is still actively working on completing implementing this scenario. This means for now that it might already work for you, but specific scenarios, such as when you're using taxonomy driven user profile fields, might not work yet. Keep this in mind when switching to the Azure AD / Entra ID option. Test everything thoroughly. If it works for your use case, you're good to go, if it doesn't yet, bear a little longer with us, it's still work in progress.

**Comment by bpw320** [comment]
I got the error: the current user has insufficient permissions to perform this operation for:
SetMultiValuedProfileProperty 
SetSingleValueProfileProperty    for property:  SPS-Location


 However, SetSingleValueProfileProperty  does work  for other  custom properties 


I use PNP.framework, set ClientContext using new AuthenticationManager().GetACSAppOnlyContext(siteurl, clientid, secret) 

 Thanks

**Comment by sidndis27** [comment]
Get-PnPUserProfileProperty works with -ClientSecret but not working with -Thumbprint. Getting Access Denied.

**Comment by WiljanD** [comment]
I fixed the issue with the latest PowerShell module on Powershell 7.2 with the command "Sync-PnPSharePointUserProfilesFromAzureActiveDirectory" all information is well documented: [https://pnp.github.io/powershell/cmdlets/Sync-PnPSharePointUserProfilesFromAzureActiveDirectory.html](url)

The service principal needs to have the following rights:
Graph API:
[User.Read.All]

Sharepoint API
[User.Read.All]
[Sites.FullControl.All]

Which will look like:
<img width="1219" alt="image" src="https://github.com/pnp/powershell/assets/109344602/324b31d7-0918-40f8-b080-cc1acf1f34e3">

The module also has commands for granting these rights in an easy way:
`Add-PnPAzureADServicePrincipalAppRole -Principal "<Service Principal Name>" -AppRole "User.Read.All" -BuiltInType MicrosoftGraph`
`Add-PnPAzureADServicePrincipalAppRole -Principal "<Service Principal Name>" -AppRole "Sites.FullControl.All" -BuiltInType SharePointOnline`
`Add-PnPAzureADServicePrincipalAppRole -Principal "<Service Principal Name>" -AppRole "User.Read.All" -BuiltInType SharePointOnline`

Then also grant the permissions in Sharepoint:
App permission:
`<AppPermissionRequests AllowAppOnlyPolicy="true"> <AppPermissionRequest Scope="http://sharepoint/content/tenant" Right="FullControl" /> </AppPermissionRequests>
`
I suspect that this rights package will help in granting rights also for Set-PnPUserProfileProperty, but I started using the latest module because it adds support for using System managed identities and User Managed Identities which I'm using in combination with a automation account to sync the attributes from Azure AD to SharePoint.

The script looks like this:
```
Connect-PnPOnline tentantname-admin.sharepoint.com -ManagedIdentity -UserAssignedManagedIdentityClientId "Service principal ObjectID"

$users = Get-PnPAzureADUser -select "OnPremisesSamAccountName"

Sync-PnPSharePointUserProfilesFromAzureActiveDirectory -UserProfilePropertyMapping @{"UserName"="OnPremisesSamAccountName";"CellPhone"="MobilePhone"} -Users $users -Folder "Documents"

Disconnect-PnPOnline
```

**Comment by ttd-kevinclement** [comment]
> Hi Bernard, the PnP projects are "community driven" and AFAIK don't have a current timeline for fixing this. However please note there are many ways to connect with SharePoint online using PnP under different circumstances, and many of them do work. See the [13 different examples](https://docs.microsoft.com/en-us/powershell/module/sharepoint-pnp/connect-pnponline). May take some digging to get the most promising example for your scenario working. I needed to use Example 3 because of non-interactive login. First attempts didn't work just, but then I found more details about dependencies and finally got it working:
> 
> The key is to follow this article [Granting access using SharePoint App-Only](https://docs.microsoft.com/en-us/sharepoint/dev/solution-guidance/security-apponly-azureacs), but to add the scope /sharepoint/**social**/tenant. Update the Permission Request XML as follows, and it will permit writing user profile properties.
> 
> ```
> <AppPermissionRequests AllowAppOnlyPolicy="true">
>   <AppPermissionRequest Scope="http://sharepoint/content/tenant" Right="FullControl" />
>   <AppPermissionRequest Scope="http://sharepoint/social/tenant" Right="FullControl" />
> </AppPermissionRequests>
> ```
> 
> This ACS which is retired now but [_explicitly still supported for SharePoint use_](https://docs.microsoft.com/en-us/sharepoint/dev/solution-guidance/security-apponly-azureacs)

Is there a way to assign these permissions using the PnP commandlets? Or some non SharePoint registration?

**Comment by AndersRask** [comment]
In my testing the issue with ACL is now solved, and so legacy auth using appreg.aspx is no longer needed.
I have tested using both system managed identity and app reg with cert. Both works fine updating user profile properties including term set fields with only API permissions

`Set-PnPUserProfileProperty -Account adelev@contoso.onmicrosoft.com -PropertyName "SPS-Location" -Value "Stockholm"`

You can ignore the exchange perm as it was only used for testing
![image](https://github.com/pnp/powershell/assets/1651150/bd3a9648-4a31-40be-a979-78bb8252abfc)

**Comment by ttd-kevinclement** [comment]
> In my testing the issue with ACL is now solved, and so legacy auth using appreg.aspx is no longer needed. I have tested using both system managed identity and app reg with cert. Both works fine updating user profile properties including term set fields with only API permissions
> 
> `Set-PnPUserProfileProperty -Account adelev@contoso.onmicrosoft.com -PropertyName "SPS-Location" -Value "Stockholm"`
> 
> The key is adding _TermStore.ReadWrite.All._ You can ignore the exchange perm as it was only used for testing ![image](https://private-user-images.githubusercontent.com/1651150/338939021-bd3a9648-4a31-40be-a979-78bb8252abfc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgyMTE1NTUsIm5iZiI6MTcxODIxMTI1NSwicGF0aCI6Ii8xNjUxMTUwLzMzODkzOTAyMS1iZDNhOTY0OC00YTMxLTQwYmUtYTk3OS03OGJiODI1MmFiZmMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYxMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MTJUMTY1NDE1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MGYzMzQ2NGM5ZjEwZTk4ZWVjY2QzYzI1NjNkYjdmNjNmNTVjOGYzMTQyOGNjNzY3NjljYmJhMDJhZWQ4ZDUzNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.htwTo4_AivKrZUCR1U2G9WCR8SY0kg4flkwK2VJHds4)

Looks like TermStore permission is all that was missing. Rest of the commands seem to work now.

**Comment by KoenZomers** [comment]
Thanks @AndersRask for sharing and @ttd-kevinclement for confirming. I will update our documentation with this great info.

I guess with that, we can also finally close this long running open issue. Thanks for providing all the feedback and thoughts here everyone.

**Comment by DavorSukeri** [comment]
Hi,

If I'm not mistaken the original issue is still not fixed?

I see something was solved concerning ACL, but the original issue was **without using ACL** and since it is going away I'm not going to use it.

So I'm now in the same situation as the OP.

Trying to read certain extension attributes from EntraID profiles and write the values into SharePoint User Profile custom properties. 

Here's what I have tried:

# Connect to Microsoft Graph API
Connect-MgGraph -Scopes "User.Read"

# Fetch users where Company Name is "CompanyXYZ"
$users = Get-MgUser -Filter "companyName eq 'CompanyXYZ'"

# Connect to SharePoint Online
Connect-PnPOnline -Url "https://companyxyz-admin.sharepoint.com" -ClientId <PnPPowerShellClientIDhere>

# Update SharePoint user profiles
foreach ($user in $users) {
    $customAttribute = $user.companyName
    Set-PnPUserProfileProperty -Account $user.UserPrincipalName -PropertyName "CompanyName" -Value $customAttribute
}


Set-PnPUserProfileProperty gives error: "This operation requires you to be managing your own data or have administrator privileges."

I have set these (Delegated) API permissions to PnP Powershell and I'm a Global Admin:

Microsoft Graph: User.Read.All
SharePoint: AllSites.FullControl, TermStore.ReadWrite.All, User.ReadWrite.All

**Comment by DavorSukeri** [comment]
> [@DavorSukeri](https://github.com/DavorSukeri) I use:
> 
> ```
> #write-host "Enter password:"
> #read-host -assecurestring | convertfrom-securestring | out-file cred.txt
> 
> $AdminSiteURL = "https://companyxyz-admin.sharepoint.com"
> 
> $UserName = "xxxx@xxx.xxx" # Tenant Administrator username
> $Password = get-content cred.txt | convertto-securestring
> $credential = New-Object System.Management.Automation.PsCredential($UserName,$password)
> 
> Connect-PnPOnline -Url $AdminSiteURL -Credentials $credential -SPOManagementShell
> ```

Thanks for you suggestion, but I don't think that would help me since eventually I would run this with Managed Identity from Azure Runbook. :(
